### ISSUE TITLE
Can't reproduce the result of ESPnet1 in ESPnet2

### ISSUE THREAD

**Issue by supernaiter** [espnet/espnet#4571]
Thank you for ESPnet team's continuous support. I have been using ESPnet and ESPnet2 for silent speech recogntion tasks for two years. (https://github.com/espnet/espnet/issues/1926)

I did visual speech recognition task using ESPnet1 approximately one year ago and got a decent 76% Character Corr. Currently I moved to ESPnet2 environment and have been failing to reproduce the equivalent result. 
Training details below:

Task: Silent Speech Recognition (using ASR pipeline)
Dataset: 1164 samples in total, 100 for testing, 100 for validation. 512 dimension representation at 30 Hz. A single sample includes 3-5 seconds resulting in 100-200 frames. The same dataset is used for both pipelines.

Config: In the ESPnet1 projects, we have been using VoxForge recipe since the dataset is small.

ESPnet1 experiment's config:

```
# espnet/egs/timit_ssc/PROJECT/conf/train.yaml
# network architecture
# encoder related
transformer-input-layer: conv2d
elayers: 12
eunits: 2048
# decoder related
dlayers: 1
dunits: 1024
# attention related
adim: 256
aheads: 4
# transformer related
model-module: "espnet.nets.pytorch_backend.e2e_asr_transformer:E2E"

# hybrid CTC/attention
mtlalpha: 0.3

# label smoothing
lsm-weight: 0.1

# minibatch related
batch-size: 24
maxlen-in: 512  # if input length  > maxlen_in, batchsize is automatically reduced
maxlen-out: 150 # if output length > maxlen_out, batchsize is automatically reduced

# optimization related
sortagrad: 0 # Feed samples from shortest to longest ; -1: enabled for all epochs, 0: disabled, other: enabled for 'other' epochs
opt: noam
epochs: 1000
dropout-rate: 0.1
accum-grad: 2
grad-clip: 5
patience: 0
transformer-lr: 20.0
transformer-warmup-steps: 25000
transformer-attn-dropout-rate: 0.0
transformer-length-normalized-loss: False
transformer-init: pytorch
```

ESPnet1 training result:

<img width="771" alt="image" src="https://user-images.githubusercontent.com/14064383/184533031-4d1ead7e-6dd1-497c-b431-becec0f6ea15.png">

ESPnet2 experiment's config:
```
# exp/asr_test1/config.yaml
config: conf/train_asr_transformer.yaml
print_config: false
log_level: INFO
dry_run: false
iterator_type: sequence
output_dir: /home/hdd/lip_visual/220812_camera_speller_512_from_espnet1/exp/asr_220812_camera_speller_512_from_espnet1
ngpu: 1
seed: 0
num_workers: 1
num_att_plot: 3
dist_backend: nccl
dist_init_method: env://
dist_world_size: null
dist_rank: null
local_rank: 0
dist_master_addr: null
dist_master_port: null
dist_launcher: null
multiprocessing_distributed: false
unused_parameters: false
sharded_ddp: false
cudnn_enabled: true
cudnn_benchmark: false
cudnn_deterministic: true
collect_stats: false
write_collected_feats: false
max_epoch: 500
patience: null
val_scheduler_criterion:
- valid
- loss
early_stopping_criterion:
- valid
- loss
- min
best_model_criterion:
-   - valid
    - acc
    - max
keep_nbest_models: 10
nbest_averaging_interval: 0
grad_clip: 5.0
grad_clip_type: 2.0
grad_noise: false
accum_grad: 8
no_forward_run: false
resume: true
train_dtype: float32
use_amp: false
log_interval: null
use_matplotlib: true
use_tensorboard: true
use_wandb: false
wandb_project: null
wandb_id: null
wandb_entity: null
wandb_name: null
wandb_model_log_interval: -1
detect_anomaly: false
pretrain_path: null
init_param: []
ignore_init_mismatch: false
freeze_param: []
num_iters_per_epoch: null
batch_size: 16
valid_batch_size: null
batch_bins: 1000000
valid_batch_bins: null
train_shape_file:
- /home/hdd/lip_visual/220812_camera_speller_512_from_espnet1/exp/asr_stats_fbank_pitch_en_char/train/speech_shape
- /home/hdd/lip_visual/220812_camera_speller_512_from_espnet1/exp/asr_stats_fbank_pitch_en_char/train/text_shape.char
valid_shape_file:
- /home/hdd/lip_visual/220812_camera_speller_512_from_espnet1/exp/asr_stats_fbank_pitch_en_char/valid/speech_shape
- /home/hdd/lip_visual/220812_camera_speller_512_from_espnet1/exp/asr_stats_fbank_pitch_en_char/valid/text_shape.char
batch_type: folded
valid_batch_type: null
fold_length:
- 800
- 150
sort_in_batch: descending
sort_batch: descending
multiple_iterator: false
chunk_length: 500
chunk_shift_ratio: 0.5
num_cache_chunks: 1024
train_data_path_and_name_and_type:
-   - dump/fbank_pitch/train/feats.scp
    - speech
    - kaldi_ark
-   - dump/fbank_pitch/train/text
    - text
    - text
valid_data_path_and_name_and_type:
-   - dump/fbank_pitch/val/feats.scp
    - speech
    - kaldi_ark
-   - dump/fbank_pitch/val/text
    - text
    - text
allow_variable_data_keys: false
max_cache_size: 0.0
max_cache_fd: 32
valid_max_cache_size: null
optim: adam
optim_conf:
    lr: 0.005
scheduler: warmuplr
scheduler_conf:
    warmup_steps: 30000
token_list:
- <blank>
- <unk>
- e
- i
- r
- s
- a
- t
- n
- o
- l
- c
- d
- u
- p
- g
- h
- m
- y
- b
- f
- w
- v
- k
- x
- j
- q
- z
- <sos/eos>
init: xavier_uniform
input_size: 512
ctc_conf:
    dropout_rate: 0.0
    ctc_type: builtin
    reduce: true
    ignore_nan_grad: null
    zero_infinity: true
joint_net_conf: null
use_preprocessor: true
token_type: char
bpemodel: null
non_linguistic_symbols: null
cleaner: null
g2p: null
speech_volume_normalize: null
rir_scp: null
rir_apply_prob: 1.0
noise_scp: null
noise_apply_prob: 1.0
noise_db_range: '13_15'
short_noise_thres: 0.5
frontend: null
frontend_conf: {}
specaug: specaug
specaug_conf:
    apply_time_warp: true
    time_warp_window: 5
    time_warp_mode: bicubic
    apply_freq_mask: true
    freq_mask_width_range:
    - 0
    - 124
    num_freq_mask: 2
    apply_time_mask: true
    time_mask_width_range:
    - 0
    - 40
    num_time_mask: 2
normalize: global_mvn
normalize_conf:
    stats_file: /home/hdd/lip_visual/220812_camera_speller_512_from_espnet1/exp/asr_stats_fbank_pitch_en_char/train/feats_stats.npz
model: espnet
model_conf:
    ctc_weight: 0.3
    lsm_weight: 0.1
    length_normalized_loss: false
preencoder: null
preencoder_conf: {}
encoder: transformer
encoder_conf:
    output_size: 256
    attention_heads: 4
    linear_units: 2048
    num_blocks: 12
    dropout_rate: 0.1
    positional_dropout_rate: 0.1
    attention_dropout_rate: 0.0
    input_layer: linear
    normalize_before: true
postencoder: null
postencoder_conf: {}
decoder: transformer
decoder_conf:
    attention_heads: 4
    linear_units: 2048
    num_blocks: 6
    dropout_rate: 0.1
    positional_dropout_rate: 0.1
    self_attention_dropout_rate: 0.0
    src_attention_dropout_rate: 0.0
required:
- output_dir
- token_list
version: '202205'
distributed: false
```

ESPnet2 training result:
<img width="556" alt="image" src="https://user-images.githubusercontent.com/14064383/184532986-766026dc-18e2-4891-938b-42e74fe9dd84.png">

**Comment by sw005320** [comment]
- Can you show the contents of the config file instead of all config information in espnet2?
- We should disable specaug.
- accum grad should be in the same range.
- decoder seems to be too big compared with espnet1

**Comment by supernaiter** [comment]
Do you mean this folder?
<img width="210" alt="image" src="https://user-images.githubusercontent.com/14064383/184677678-24d66e20-e53f-4669-80ec-24e4b96c56a8.png">

**Comment by sw005320** [comment]
It would be a yaml file that is used in run.sh.

**Comment by supernaiter** [comment]
I'm afraid I couldn't get what you really need, I uploaded projects in: https://github.com/supernaiter/cameraspeller_egs2

We really appreciate your support.

**Comment by b-flo** [comment]
He was talking about [this line in run.sh](https://github.com/supernaiter/cameraspeller_egs2/blob/main/run.sh#L24) which means you used [this yaml config](https://github.com/supernaiter/cameraspeller_egs2/blob/main/conf/train_asr_transformer.yaml) for your training.

However, it's not equivalent to what is reported here. For example,  `num_blocks` and `linear_units` in `decoder_conf` are different across configs.

Nevertheless, the ESPnet2 config in your repo seems somewhat equivalent to the ESPnet1 except for what was pointed out by @sw005320:

- Disable specaug.
- Set equivalent parameters for the decoder part.
- Use the same `accum_grad` (`lr` is in range but may be adjusted).
- Lower `warmup_steps` (`30000`-> `25000`).

**Comment by supernaiter** [comment]
Thank you @sw005320 -san and @b-flo -san, I am running the experiments following your instructions.

**Comment by supernaiter** [comment]
<img width="628" alt="image" src="https://user-images.githubusercontent.com/14064383/184854943-da88d5cd-12e1-4ab0-af3b-52ae6ee55d05.png">

a quick report
epoch=105

**Comment by sw005320** [comment]
It's getting better.
I'd suggest reducing the learning rate.

**Comment by sw005320** [comment]
specaugment makes the debug difficult.
Now, it seems that training is working but over-training happens.
So, can we enable specaugment?
You may need to tune the learning rate when you change the setup.
I hope it will reach around 70% (espnet1 results).

**Comment by supernaiter** [comment]
Ok now I am running with SpecAugment. Thank you.
Do you have any recommended confs for training using small dataset?

**Comment by supernaiter** [comment]
Quick report: training with SpecAugment doesn't seem working well.
<img width="614" alt="image" src="https://user-images.githubusercontent.com/14064383/184930915-83401f58-e078-41b8-ac6d-a044d3cf4b18.png">

**Comment by sw005320** [comment]
I would recommend you not conclude it without tuning the learning rate and so on.

**Comment by supernaiter** [comment]
We have been testing on many different settings. 
So far, this asr config made the best result.
`batch_type: folded
batch_size: 24
accum_grad: 2
max_epoch: 5000
patience: none
# The initialization method for model parameters
init: xavier_uniform
best_model_criterion:
-   - valid
    - acc
    - max
keep_nbest_models: 10


encoder: transformer
encoder_conf:
    output_size: 256
    attention_heads: 4
    linear_units: 2048
    num_blocks: 3
    dropout_rate: 0.5
    positional_dropout_rate: 0.5
    attention_dropout_rate: 0.5
    input_layer: linear
    normalize_before: true

decoder: transformer
decoder_conf:
    attention_heads: 4
    linear_units: 1024
    num_blocks: 1
    dropout_rate: 0.5
    positional_dropout_rate: 0.5
    self_attention_dropout_rate: 0.5
    src_attention_dropout_rate: 0.5

model_conf:
    ctc_weight: 0.3
    lsm_weight: 0.1
    length_normalized_loss: false


optim: adam
optim_conf:
    lr: 0.0005
    #lr: 1.0

    
scheduler: warmuplr
scheduler_conf:
    warmup_steps: 25000`

<img width="438" alt="image" src="https://user-images.githubusercontent.com/14064383/185615428-2cbf7d5f-4813-440c-8f15-9e40fad6f9b7.png">



We are still struggling in making SpecAugment work well.

`specaug: specaug
specaug_conf:
    apply_time_warp: true
    time_warp_window: 5
    time_warp_mode: bicubic
    apply_freq_mask: false
    freq_mask_width_range:
    - 0
    - 20
    num_freq_mask: 2
    apply_time_mask: true
    time_mask_width_range:
    - 0
    - 5
    num_time_mask: 2


batch_type: folded
batch_size: 24
accum_grad: 2
max_epoch: 5000
patience: none
# The initialization method for model parameters
init: xavier_uniform
best_model_criterion:
-   - valid
    - acc
    - max
keep_nbest_models: 10


encoder: transformer
encoder_conf:
    output_size: 256
    attention_heads: 4
    linear_units: 2048
    num_blocks: 3
    dropout_rate: 0.5
    positional_dropout_rate: 0.5
    attention_dropout_rate: 0.5
    input_layer: linear
    normalize_before: true

decoder: transformer
decoder_conf:
    attention_heads: 4
    linear_units: 1024
    num_blocks: 1
    dropout_rate: 0.5
    positional_dropout_rate: 0.5
    self_attention_dropout_rate: 0.5
    src_attention_dropout_rate: 0.5

model_conf:
    ctc_weight: 0.3
    lsm_weight: 0.1
    length_normalized_loss: false


optim: adam
optim_conf:
    lr: 0.0001
    #lr: 1.0

    
scheduler: warmuplr
scheduler_conf:
    warmup_steps: 25000

`

<img width="383" alt="image" src="https://user-images.githubusercontent.com/14064383/185616526-396dd3b1-1a5a-4441-a440-c2723afe0fc7.png">

**Comment by sw005320** [comment]
Thanks for sharing.
I have two comments
- The dropout rate seems to be too large (but not sure. It might be OK).
- I feel that the number of samples is too short to do (silent) speech recognition and may need some more modifications in the front-end part (currently linear, which sounds correct to me, though).

**Comment by supernaiter** [comment]
Thank you for your advice.

- I'll lower the dropout rates.
- I'll try different settings in the front-end parts.

But I'm still wondering why SpecAugment is not working as well as ESPnet.

**Comment by sw005320** [comment]
Can you tell me why you changed the SpecAugment configurations from default?
You seem to disable `apply_freq_mask`
Also, `time_mask_width_range` values are very different from default.
